### ISSUE TITLE
Authentication and Registration when user goes directly to che.openshift.io

### ISSUE THREAD

**Issue by slemeur** [openshiftio/openshift.io#3814]
## Description
To support the transition of codenvy.io to che.openshift.io, we need to handle the different authentication flow, when a user will access directly to che.openshift.io.

## Proposed Flows
**User, non-authenticated + not having osio account, goes to che.openshift.io**
- Authentication page displayed
  - The page is today displaying only “OpenShift.io” information
  - Could we specialize the authentication page, so the user can get information that he is signing up for OpenShift.io, but will get access to cloud workspaces with Che? 
- User does not have an account, he has to create a new account
  - Similarly, the page to create a new account is showing only “OpenShift.io” information
  - Could we specialize the authentication page, so the user can get information that he is signing up for OpenShift.io, but will get access to cloud workspaces with Che? 
  - Sign-up emails, are specialized for “OpenShift.io” and are not including Che information. 
  - Could we have a separate email, when a user is signing up in the che.openshift.io flow
- After creating the account, the user should be redirected into his dashboard on che.openshift.io

**User, non-authenticated + having osio account, goes to che.openshift.io**
- Authentication page displayed
  - The page is today displaying only “OpenShift.io” information
  - Could we specialize the authentication page, so the user can get information that he is signing up for OpenShift.io, but will get access to cloud workspaces with Che? 
- User follow the flow to authenticate
- After authenticating, the user is redirected into his dashboard on che.openshift.io (already working)

**User, authenticated, goes to che.openshift.io**
- The user is redirected into his dashboard on che.openshift.io (already working)

**Comment by joshuawilson** [comment]
which team will work on this? Che or Auth or both?

**Comment by slemeur** [comment]
@ibuziuk : 
The pages you put the screenshots for are the right ones. You are right, we want to rebrand these pages but not change the registration flow. 

The rebranding should:
- Change the title
  - instead of "Register for the OpenShift.io" it should be something like "Register to Eclipse Che Live" (consider `Eclipse Che Live` as a placeholder for now)
  - we would add a small subtitle: "Eclipse Che online instance powered by OpenShift.io"
- Change the logo in the header

**Comment by ibuziuk** [comment]
@slemeur thanks for clarification, so for me it sounds that the issue for UI / UX team. Basically, the issue is about applying different css themes based on `document.referrer` [1] e.g. if it "https://rhche.openshift.io/dashboard/" / "https://che.openshift.io/dashboard/" new che theme should be applied on registration pages. I guess, no input is required from Che / Auth team regarding this issue.

[1] https://developer.mozilla.org/en-US/docs/Web/API/Document/referrer

**Comment by ibuziuk** [comment]
@slemeur Joshua have added `team/che/osio` label but I believe it is pure UI / UX thingy, and `team/ux` would be more appropriate here, right ?

**Comment by joshuawilson** [comment]
@ibuziuk @slemeur you can pass it along to whatever team is working the issue. I added this one since you are writing on it. 
cc @openshiftio/uxd-team

**Comment by slemeur** [comment]
@catrobson : Could your team handle this issue ?

**Comment by catrobson** [comment]
@slemeur yes, thanks for the ping as I missed this earlier.  @mindreeper2420 could you work with Tiffany to get these changes into the SSO pages for Che.openshift.io branding and information?

@slemeur is there an "approved" name and logo for this already, or do we need to work with Brand on getting a new logo made?

**Comment by slemeur** [comment]
I've replied over email in a separate thread, but there is no new logo or name that we want to use here. 

The users are going to be redirected to a "pure Eclipse Che" experience. We can reuse Che logo in the header. We can also have a subtitle "Powered by Red Hat on OpenShift.IO"

**Comment by catrobson** [comment]
@davidfestal - Many of the things you pointed out are being discussed here:  https://docs.google.com/presentation/d/1P8p_6Va1hExB3y7vZvmAxkHjliDpzOZVDaT6ox-2lG8/edit#slide=id.g3e4795d877_0_20

@sbose78 FYI

**Comment by davidfestal** [comment]
@catrobson yes, Thanks for the information. I already added some comments in this document. In fact it happens that the google doc has been created / shared while I was writing the comment above: I found it just after.

**Comment by ibuziuk** [comment]
@davidfestal thanks for a detailed analyses / explanations and the demo - https://youtu.be/fGZZZVY93Uc

> What I wonder is: shouldn't we simply provide a sort of facade / proxy small application (mainly some client-side Javascript code) that would chain these steps as necessary, waiting for the intermediate asynchronous tasks to finish (provisioning and, later, tenant init), before redirecting to the real Che server URL initially requested ?

Am I correct that this would be a new app hosted on osd ? I personally like this idea more than the other proposal [1]
[1] https://docs.google.com/presentation/d/1P8p_6Va1hExB3y7vZvmAxkHjliDpzOZVDaT6ox-2lG8/edit#slide=id.g3e4795d877_0_75

@slemeur @l0rd improving the situation with fixing the registration flow would definitely require more  time to fix, which would delay the Phase 1 of migration to `che.openshift.io` even more. Could we consider explicitly stating that the RHD account is required in order to get started on https://www.eclipse.org/che/getting-started/ similar to `codenvy.io` account info:

![image](https://user-images.githubusercontent.com/1461122/43203056-718d547e-901d-11e8-947e-27f371e042d2.png)

**Comment by davidfestal** [comment]
>>    What I wonder is: shouldn't we simply provide a sort of facade / proxy small application (mainly some client-side Javascript code) that would chain these steps as necessary, waiting for the intermediate asynchronous tasks to finish (provisioning and, later, tenant init), before redirecting to the real Che server URL initially requested ?
>
> Am I correct that this would be a new app hosted on osd ? I personally like this idea more than the other proposal [1]

yes, or there could be an even simpler solution implemented fully in rh-che by overriding the [`che.keycloak.js_adapter_url`](https://github.com/eclipse/che/blob/master/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties#L147) in order to use a customized version of the Keycloak client script (that would delegate to the standard one, but simply add the provisionning + init tenant steps automatically when necessary after the initial login).

That would make things fully transparent.

I'm currently testing this fully locally in a test html page (since it's pure client-side javascript). And if it works locally, it would be pretty easy to test it on dev-cluster by simply overriding the `che.keycloak.js_adapter_url` property to the new script available on a GitHub branch.

Do you want me to open an new issue in rh-che for the POC work ?

**Comment by ibuziuk** [comment]
> Do you want me to open an new issue in rh-che for the POC work ?
@davidfestal yes, please. We discussed it with @slemeur and he confirmed that we could spend  time on POC that would provide better UX for registrations via `che.openshift.io` in this sprint.

**Comment by ruthetdav** [comment]
@ibuziuk @slemeur great !

**Comment by davidfestal** [comment]
@slemeur @ibuziuk great !

**Comment by ibuziuk** [comment]
@slemeur @mindreeper2420 is any work still planned to re-branding of the registration pages ?

**Comment by ibuziuk** [comment]
@slemeur @l0rd FYI confirmation emails still have openshift.io branding for e2e che.openshift.io registration flow:

![image](https://user-images.githubusercontent.com/1461122/53918709-fe135600-4067-11e9-91f6-04ad4d84e8b3.png)
![image](https://user-images.githubusercontent.com/1461122/53918734-0e2b3580-4068-11e9-847c-f82fd4ff2501.png)
